<a href="https://colab.research.google.com/github/daryafralova/CodeTasks/blob/main/Pandas_%D0%9F%D0%BB%D0%B0%D1%82%D0%B5%D0%B6%D0%B5%D1%81%D0%BF%D0%BE%D1%81%D0%BE%D0%B1%D0%BD%D0%BE%D1%81%D1%82%D1%8C_%D0%BA%D0%BB%D0%B8%D0%B5%D0%BD%D1%82%D0%BE%D0%B2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Название датасета:** Надежность заёмщиков - исследование платёжеспособности

**Цель:** Определить влияние различных характеристик клиентов на вероятность своевременного погашения кредита ответив на следующие вопросы:

1.   Влияет ли количество детей на своевременный возврат кредита?
2.   Влияет ли семейное положение на своевременный возврат кредита?
3.   Влияет ли уровень дохода на своевременный возврат кредита?
4.   Влияют ли цели кредита на его своевременный возврат?







**Ход иследования**

1.   Обзор данных
2.   Предобработка данных
3.   Ответы на вопросы исследования
4.   Общий вывод





In [ ]:
import gdown
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

###Загрузка и обзор данных


*   Проверка на наличие ошибок
*   Оценка качества данных и влияние найденных ошибок на качество  исследования



In [ ]:
gdown.download('https://drive.google.com/uc?id=14zOYNZI7w7TvLVZIhJLmoustW04Bznti')

Downloading...
From: https://drive.google.com/uc?id=14zOYNZI7w7TvLVZIhJLmoustW04Bznti
To: /content/data.csv
100%|██████████| 3.32M/3.32M [00:00<00:00, 26.9MB/s]


'data.csv'

In [ ]:
data = pd.read_csv('/content/data.csv', sep=',')

In [ ]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

*   children — количество детей в семье
*   days_employed — общий трудовой стаж
*   dob_years — возраст клиента
*   education — уровень образования клиента
*   education_id — идентификатор уровня образования
*   family_status — семейное положение
*   family_status_id — идентификатор семейного положения
*   gender — пол клиента
*   income_type — тип занятости
*   debt — имел ли задолженность по возврату кредитов
*   total_income — ежемесячный доход
*   purpose — цель получения кредита



In [ ]:
#Вывод типы данных всех столбцов
display(data.dtypes)

children              int64
days_employed       float64
dob_years             int64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
income_type          object
debt                  int64
total_income        float64
purpose              object
dtype: object

In [ ]:
#смотрим есть ли NaN значения
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [ ]:
#Смотрим кол-во дубликатов
data.duplicated().value_counts()

False    21471
True        54
Name: count, dtype: int64

In [ ]:
# вывод показателей
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

По первичному обзору данных выявили следующие аномалии по следующим столбцам:
*   children- кол-во детей не может быть отрицательным -1.000000
*   children- кол-во детей в кол-ве 20.000000 тоже выглядит подозрительным
*   days_employed-Общий трудовой стаж также выглядит подозрительно так как есть и слишком большие значения и отрицательные тоже
*   dob_years - возраст также не может быть равен 0
*   total_income - нужно проверить на выбросы
*   Также в данных присутствуют дубликаты, которые следует удалить
*   Найдены данные с NaN значениями которые нужно будет обработать





###Предварительная обработка данных


*   Удаление дубликатов
*   Обработка пропущенных значений
*   Обработка аномалий по столбцам : children, dob_years
*   Преобразование типов данных
*   Обзор данных по остальным столбцам
*   Проверка данных после предварительной обработки







**Удаляем дубликаты**

In [ ]:
# удаление явных дубликатов (с удалением старых индексов и формированием новых)
data = data.drop_duplicates().reset_index(drop=True)
data.duplicated().sum()

0

**Обработка пропущенных значений**

В двух столбцах есть пропущенные значения. Один из столбцов с пропущенными значениями — total_income — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца income_type.

**Почему берём медиану, а не среднее?**

*   Медиана устойчива к выбросам. Если у одного человека в группе очень высокий доход (например, миллионы), среднее значение может сильно исказиться
*   Среднее значение чувствительно к аномалиям, тогда как медиана более точно отражает типичное значение дохода.



In [ ]:
for t in data['income_type'].unique():#получаем список уникальных значений в столбце income_type,  for перебирает каждое из этих значений
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \#income_type == i- выбираем строки с конкретным типом дохода + фильтруем строки где total_income пустой (isna())
    data.loc[(data['income_type'] == t), 'total_income'].median() #вычисляется медианное значение столбца total_income

In [ ]:
data.isna().sum() #проверяем результат

children               0
days_employed       2120
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dtype: int64

Один из столбцов с пропущенными значениями — days_employed — хранит данные о стаже.Заполним пропуски в нём медианными значениями по типу занятости income_type

Также изменим отрицательные значения(стаж не может быть отрицательным, это аномальные значения)

In [ ]:
# возьмем модуль значений с помощью abs()
data['days_employed'] = data['days_employed'].abs()

In [ ]:
# выведем медианное значение трудового стажа в днях
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

**Заполним пропуски в столбце days_employed медианными значениями по типу занятости income_type.**

In [ ]:
for i in data['income_type'].unique(): #получаем список уникальных значений в столбце income_type,  for перебирает каждое из этих значений
    data.loc[(data['income_type'] == i) & (data['days_employed'].isna()), 'days_employed'] = \#income_type == i- выбираем строки с конкретным типом дохода + фильтруем строки где days_employed пустой (isna())
    data.loc[(data['income_type'] == i), 'days_employed'].median() #заменяем строки где days_employed пустой медианным значением

In [ ]:
data.isna().sum()  #проверяем результат

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

**Столбец children**

In [ ]:
# Подсчет количества уникальных значений в столбце 'children' DataFrame
data['children'].value_counts()

children
 0     14107
 1      4809
 2      2052
 3       330
 20       76
-1        47
 4        41
 5         9
Name: count, dtype: int64

In [ ]:
#Удаляем аномальные значения из children
data = data[(data['children'] != -1) & (data['children'] != 20)] #создаёт булевый массив (True, False), где True означает, что значение в столбце children не равно -1
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

**Cтолбец dob_years**

In [ ]:
# Посмотрим количество строк в датафрейме, где возраст меньше 18 лет(минимальны возраст заёмщика)
data.loc[data['dob_years'] < 18, 'dob_years'].count()

100

In [ ]:
#Посмотрим, являются ли строки с возрастом менее 18 лет - нулём. Для этого посчитаем кол-во строк меньше 18 лет и не равных 0.
data.loc[(data['dob_years'] < 18) & (data['dob_years'] != 0), 'dob_years'].count()

0

In [ ]:
# Пройдемся циклом  по уникальным значениям колонки income_type
# Найдем среднее значение
# Заполним пропуски полученными значениями
for t in data['income_type'].unique():
    mean_value = data.loc[data['income_type'] == t, 'dob_years'].mean()
    data.loc[(data['dob_years'] == 0) & (data['income_type'] == t), 'dob_years'] = mean_value

<ipython-input-81-87a70e72ec72>:6: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '39.81155870078026' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  data.loc[(data['dob_years'] == 0) & (data['income_type'] == t), 'dob_years'] = mean_value


In [ ]:
data.loc[data['dob_years'] == 0, 'dob_years'].count()

0

**Преобразование типов данных**

In [ ]:
data['total_income'] = data['total_income'].astype(int) # меняем вещественные значения на целочисленные
data['days_employed'] = data['days_employed'].astype(int) # меняем вещественные значения на целочисленные
data['education'] = data['education'].str.lower() # приводим к нижнему регистру столбец education

**Обзор данных по остальным столбцам**

In [ ]:
# Смотрим на значения столбца education
data['education'].sort_values().value_counts()

education
среднее                15091
высшее                  5228
неоконченное высшее      741
начальное                282
ученая степень             6
Name: count, dtype: int64

In [ ]:
# Смотрим на значения столбца gender
data['gender'].sort_values().value_counts()

gender
F      14107
M       7240
XNA        1
Name: count, dtype: int64

In [ ]:
# Посмотрим на строку с некорректным значением в столбце gender
data.loc[data['gender'] == 'XNA']

children  days_employed  dob_years            education  education_id  \
10690         0           2358       24.0  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10690  гражданский брак                 1    XNA   компаньон     0   

       total_income               purpose  
10690        203905  покупка недвижимости

Вывод по столбцу gender: определить пол не представляется возможным.

Т.к. столбец gender для исследований нам не понадобится оставим это значение как есть, чтобы не нарушать данные.

**Проверка данных после предварительной обработки**

In [ ]:
display(data)

children  days_employed  dob_years education  education_id  \
0             1           8437       42.0    высшее             0   
1             1           4024       36.0   среднее             1   
2             0           5623       33.0   среднее             1   
3             3           4124       32.0   среднее             1   
4             0         340266       53.0   среднее             1   
...         ...            ...        ...       ...           ...   
21466         1           4529       43.0   среднее             1   
21467         0         343937       67.0   среднее             1   
21468         1           2113       38.0   среднее             1   
21469         3           3112       38.0   среднее             1   
21470         2           1984       40.0   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21466  гражданский брак                 1      F   компаньон     0   
21467   женат / замужем                 0      F   пенсионер     0   
21468  гражданский брак                 1      M   сотрудник     1   
21469   женат / замужем                 0      M   сотрудник     1   
21470   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  
0            253875                 покупка жилья  
1            112080       приобретение автомобиля  
2            145885                 покупка жилья  
3            267628    дополнительное образование  
4            158616               сыграть свадьбу  
...             ...                           ...  
21466        224791             операции с жильем  
21467        155999          сделка с автомобилем  
21468         89672                  недвижимость  
21469        244093  на покупку своего автомобиля  
21470         82047         на покупку автомобиля  

[21348 rows x 12 columns]

In [ ]:
data.info() # методом info проверим тип переменных в колонках

<class 'pandas.core.frame.DataFrame'>
Index: 21348 entries, 0 to 21470
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21348 non-null  int64  
 1   days_employed     21348 non-null  int64  
 2   dob_years         21348 non-null  float64
 3   education         21348 non-null  object 
 4   education_id      21348 non-null  int64  
 5   family_status     21348 non-null  object 
 6   family_status_id  21348 non-null  int64  
 7   gender            21348 non-null  object 
 8   income_type       21348 non-null  object 
 9   debt              21348 non-null  int64  
 10  total_income      21348 non-null  int64  
 11  purpose           21348 non-null  object 
dtypes: float64(1), int64(6), object(5)
memory usage: 2.1+ MB


In [ ]:
# вывод показателей
data.describe()

children  days_employed     dob_years  education_id  \
count  21348.000000   21348.000000  21348.000000  21348.000000   
mean       0.473674   67231.124461     43.490490      0.817079   
std        0.752199  139348.906185     12.234565      0.549115   
min        0.000000      24.000000     19.000000      0.000000   
25%        0.000000    1025.000000     33.000000      1.000000   
50%        0.000000    1996.000000     43.000000      1.000000   
75%        1.000000    5347.500000     53.000000      1.000000   
max        5.000000  401755.000000     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21348.000000  21348.000000  2.134800e+04  
mean           0.974283      0.081132  1.653184e+05  
std            1.421299      0.273044  9.827990e+04  
min            0.000000      0.000000  2.066700e+04  
25%            0.000000      0.000000  1.075595e+05  
50%            0.000000      0.000000  1.425940e+05  
75%            1.000000      0.000000  1.957885e+05  
max            4.000000      1.000000  2.265604e+06

**Вывод по результатам предварительной обработки данных**



*   Дубликаты удалены
*   Пропущенных значений больше нет
*   Аномалии по столбцам children и dob_years не выявлены
*   Преобразовали столбцы days_employed и total_income к типу данных int
*   По столбцу gender выявлена запись с непонятным описанием.Но так как различие по гендерному признаку не пригодиться в дальнейшем анализе данных, данную аномалию оставили без изменений, чтобы не нарушать целостность данных
*   По столбцу days_employed - Отрицательных значений нет. Минимальное значение - 24 дня. Максимальное значение 401755 стажа - аномально. Допустим, что это значение в часах, тогда получим около 16740 дней стажа(что соответствует примерно 46 годам стажа). Учитывая, что максимальный возраст клиента 75 лет, предположение, что аномальные значения приведены в часах считаю верным. Приведем все аномальные значения в столбце к значениям в днях. Т.к. миниальный возраст для оформления по трудовому договору - 16 лет, а максимальное значение возраста - 75 лет. Возьмем границу стажа равную 59 лет (округлив получим 21500 дней). Все значения выше будем считать ошибочными и разделим их на 24 часа, приведя к дням




In [ ]:
# Приводим аномальные значения в столбце days_employed к значениям в днях
data.loc[data['days_employed'] > 21500, 'days_employed'] = data.loc[data['days_employed'] > 21500, 'days_employed'] / 24

<ipython-input-90-be8da636a02f>:2: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '[14177.75       15217.20833333 16678.375      ... 15090.04166667
 15583.125      14330.70833333]' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  data.loc[data['days_employed'] > 21500, 'days_employed'] = data.loc[data['days_employed'] > 21500, 'days_employed'] / 24


In [ ]:
# Проверим
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1        8437.00       42.0    высшее             0   
1         1        4024.00       36.0   среднее             1   
2         0        5623.00       33.0   среднее             1   
3         3        4124.00       32.0   среднее             1   
4         0       14177.75       53.0   среднее             1   
5         0         926.00       27.0    высшее             0   
6         0        2879.00       43.0    высшее             0   
7         0         152.00       50.0   среднее             1   
8         2        6929.00       35.0    высшее             0   
9         0        2188.00       41.0   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [ ]:
data['days_employed'] = data['days_employed'].astype(int) # с помощью метода astype() преобразуем данные к типу int
data.head() # проверка

children  days_employed  dob_years education  education_id  \
0         1           8437       42.0    высшее             0   
1         1           4024       36.0   среднее             1   
2         0           5623       33.0   среднее             1   
3         3           4124       32.0   среднее             1   
4         0          14177       53.0   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

##Категоризация данных для дальнейшего анализа

Создадим столбец total_income_category с категориями:

*   0–30000 — 'E';
*   30001–50000 — 'D';
*   50001–200000 — 'C';
*   200001–1000000 — 'B';
*   1000001 и выше — 'A'.







In [ ]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [ ]:
data['total_income_category'] = data['total_income'].apply(categorize_income)


Сформирует новый столбец purpose_category, в который войдут следующие категории:

*   'операции с автомобилем',
*   'операции с недвижимостью',
*   'проведение свадьбы',
*   'получение образования'.








In [ ]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [ ]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

**Выведем перечень уникальных целей взятия кредита**

In [ ]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Проверяем как выглядят наши данные для дальнейшего анализа**

In [ ]:
display(data)

children  days_employed  dob_years education  education_id  \
0             1           8437       42.0    высшее             0   
1             1           4024       36.0   среднее             1   
2             0           5623       33.0   среднее             1   
3             3           4124       32.0   среднее             1   
4             0          14177       53.0   среднее             1   
...         ...            ...        ...       ...           ...   
21466         1           4529       43.0   среднее             1   
21467         0          14330       67.0   среднее             1   
21468         1           2113       38.0   среднее             1   
21469         3           3112       38.0   среднее             1   
21470         2           1984       40.0   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21466  гражданский брак                 1      F   компаньон     0   
21467   женат / замужем                 0      F   пенсионер     0   
21468  гражданский брак                 1      M   сотрудник     1   
21469   женат / замужем                 0      M   сотрудник     1   
21470   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  \
0            253875                 покупка жилья                     B   
1            112080       приобретение автомобиля                     C   
2            145885                 покупка жилья                     C   
3            267628    дополнительное образование                     B   
4            158616               сыграть свадьбу                     C   
...             ...                           ...                   ...   
21466        224791             операции с жильем                     B   
21467        155999          сделка с автомобилем                     C   
21468         89672                  недвижимость                     C   
21469        244093  на покупку своего автомобиля                     B   
21470         82047         на покупку автомобиля                     C   

               purpose_category  
0      операции с недвижимостью  
1        операции с автомобилем  
2      операции с недвижимостью  
3         получение образования  
4            проведение свадьбы  
...                         ...  
21466  операции с недвижимостью  
21467    операции с автомобилем  
21468  операции с недвижимостью  
21469    операции с автомобилем  
21470    операции с автомобилем  

[21348 rows x 14 columns]

##Анализ данных

1. Влияет ли количество детей на своевременный возврат кредита?
 Для ответа на этот вопрос сделаем следующее:
*   Cоздадим дополнительный столбец в датафрейме, в котором сделаем категоризацию по кол-ву детей
*   Cоздадим сводную таблицу.



In [ ]:
# Функция категоризации по детям
def children_category(children):
    if 1 <= children <= 2:
        return 'есть дети'
    if children >= 3:
        return 'многодетный'
    else:
        return 'нет детей'

In [ ]:
# Создаем столбец children_rank
data['children_rank'] = data['children'].apply(children_category)

In [ ]:
# Проверяем результат. Выведем последние 5 значений
data.tail()

children  days_employed  dob_years education  education_id  \
21466         1           4529       43.0   среднее             1   
21467         0          14330       67.0   среднее             1   
21468         1           2113       38.0   среднее             1   
21469         3           3112       38.0   среднее             1   
21470         2           1984       40.0   среднее             1   

          family_status  family_status_id gender income_type  debt  \
21466  гражданский брак                 1      F   компаньон     0   
21467   женат / замужем                 0      F   пенсионер     0   
21468  гражданский брак                 1      M   сотрудник     1   
21469   женат / замужем                 0      M   сотрудник     1   
21470   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  \
21466        224791             операции с жильем                     B   
21467        155999          сделка с автомобилем                     C   
21468         89672                  недвижимость                     C   
21469        244093  на покупку своего автомобиля                     B   
21470         82047         на покупку автомобиля                     C   

               purpose_category children_rank  
21466  операции с недвижимостью     есть дети  
21467    операции с автомобилем     нет детей  
21468  операции с недвижимостью     есть дети  
21469    операции с автомобилем   многодетный  
21470    операции с автомобилем     есть дети

In [ ]:
debt_children = data.groupby(['children_rank', 'debt'])['gender'].count().unstack() #группируем по children_rank и debt, и с помощью  (unstack()) преобразуем таблиуцу с 2мя столбцами
debt_children.columns = ['no_debt', 'debt'] # Переименовываем столбцы
debt_children['share_of_debtors'] = debt_children['debt'] / (debt_children['debt'] + debt_children['no_debt'])# Считаем долю должников
debt_children['share_of_debtors'] = debt_children['share_of_debtors'].map('{:.1%}'.format)# Переводим долю в проценты
debt_children = debt_children.sort_values(by='share_of_debtors', ascending=False) # Сортируем по убыванию
display(debt_children)

no_debt  debt share_of_debtors
children_rank                                
есть дети         6223   638             9.3%
многодетный        349    31             8.2%
нет детей        13044  1063             7.5%

Вывод: Наличие детей повышает вероятность задолжености. Вероятно, это связано с дополнительными расходами на детей. Однако, можно заметить, что в многодетных семьях уровень должников ниже (8.2%), чем в семьях с одним или двумя детьми (9.3%). Это может быть связано либо с тем, что существуют программы поддержки многодетных семей или количество заемщиков в категории "многодетный" для данного сравнения недостаточно. Для категории граждан без детей самый низкий показатель должников (7.3%).

2.Влияет ли семейное положение на своевременный возврат кредита?

In [ ]:
debt_family_status = data.groupby(['family_status', 'debt'])['gender'].count().unstack() #Группируем по family_status и debt
debt_family_status.columns = ['no_debt', 'debt']# Переименовываем столбцы
debt_family_status['share_of_debtors'] = debt_family_status['debt'] / (debt_family_status['debt'] + debt_family_status['no_debt'])# Считаем долю должников
debt_family_status['share_of_debtors'] = debt_family_status['share_of_debtors'].map('{:.1%}'.format)# Переводим долю в проценты
debt_family_status = debt_family_status.sort_values(by='share_of_debtors', ascending=False)# Сортируем по убыванию
display(debt_family_status)

no_debt  debt share_of_debtors
family_status                                        
Не женат / не замужем     2523   273             9.8%
гражданский брак          3761   385             9.3%
женат / замужем          11339   927             7.6%
в разводе                 1105    84             7.1%
вдовец / вдова             888    63             6.6%

Вывод: Холостые и живущие в граждансокм браке пары больше подвержены риску стать должниками. Вероятно, это связано с тем, что у них больше свободы действий по распоряжению своими средствами. У них нет совместных накоплений, каждый распоряжается своими средствами по своему. Остальные категории, побывавшие в браке или на текущей момент в браке имеют меньшие проценты должников. Наиболее низкий процент у категории вдовец/вдова (6.6%). Это может быть связано с дополнительными средствами, доставшимися по наследству.

3.Влияет ли уровень дохода на своевременный возврат кредита?

In [ ]:
debt_total_income = data.groupby(['total_income_category', 'debt'])['gender'].count().unstack() #Группируем по total_income_category и debt
debt_total_income.columns = ['no_debt', 'debt']# Переименовываем столбцы
debt_total_income['share_of_debtors'] = debt_total_income['debt'] / (debt_total_income['debt'] + debt_total_income['no_debt'])# Считаем долю должников
debt_total_income['share_of_debtors'] = debt_total_income['share_of_debtors'].map('{:.1%}'.format)# Переводим долю в проценты
debt_total_income = debt_total_income.sort_values(by='share_of_debtors', ascending=False)# Сортируем по убыванию
display(debt_total_income)

no_debt  debt share_of_debtors
total_income_category                                
E                           20     2             9.1%
C                        14585  1353             8.5%
A                           23     2             8.0%
B                         4660   354             7.1%
D                          328    21             6.0%

Вывод: Категория граждан (С) со средним доходом (50001–200000) больше подвержены риску стать должниками. Несмотря на то, что в итоговой таблицы должников данная категория граждан имеет не самый высокий процент по должникам, для точного исследования мы должны рассматривать репрезентативные выборки(категории граждан). Соответственно категории "E" и "A" мы не берем в расчет (из-за маленькой выборки), а выводы делаем опираясь на категории C, B, D. Сравнивая эти категории делаем вывод, что наибольший риск стать должниками у категории C, а наименьший у категории D.

4. Влияют ли цели кредита на его своевременный возврат?

Сформирует новый столбец purpose_category, в который войдут следующие категории:

'операции с автомобилем',
'операции с недвижимостью',
'проведение свадьбы',
'получение образования'.

In [ ]:
debt_purpose_category = data.groupby(['purpose_category', 'debt'])['gender'].count().unstack()
debt_purpose_category.columns = ['no_debt', 'debt']# Переименовываем столбцы
debt_purpose_category['share_of_debtors'] = debt_purpose_category['debt'] / (debt_purpose_category['debt'] + debt_purpose_category['no_debt'])# Считаем долю должников
debt_purpose_category['share_of_debtors'] = debt_purpose_category['share_of_debtors'].map('{:.1%}'.format)# Переводим долю в проценты
debt_purpose_category = debt_purpose_category.sort_values(by='share_of_debtors', ascending=False)# Сортируем по убыванию
display(debt_purpose_category)

no_debt  debt share_of_debtors
purpose_category                                        
операции с автомобилем       3881   400             9.3%
получение образования        3620   369             9.3%
проведение свадьбы           2141   183             7.9%
операции с недвижимостью     9974   780             7.3%

Вывод: Граждане, берущие кредиты на получение образования и операции с автомобилем больше подвержены риску стать должниками. Эти две категории имеют одинаковый процент должников (9.3%). Граждане, которые берут кредит для проведения операций с недвижимостью более отвественно подходят к его погашению.

###Общий вывод

*   Наличие детей повышает вероятность задолжености.
*   Холостые и живущие в граждансокм браке пары больше подвержены риску стать должниками.
*   Категория граждан со средним доходом (50001–200000) больше подвержены риску стать должниками.
*   Граждане, берущие кредиты на получение образования и операции с автомобилем больше подвержены риску стать должниками.



